pkg/distributor/ha_tracker_http.go

```diff
@@ -0,0 +1,100 @@
+package distributor
+
+import (
+	"html/template"
+	"net/http"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/prometheus/prometheus/pkg/timestamp"
+)
+
+const trackerTpl = `
+<!DOCTYPE html>
+<html>
+	<head>
+		<meta charset="UTF-8">
+		<title>Cortex HA Tracker Status</title>
+	</head>
+	<body>
+		<h1>Cortex HA Tracker Status</h1>
+		<p>Current time: {{ .Now }}</p>
+		<table width="100%" border="1">
+			<thead>
+				<tr>
+					<th>User ID</th>
+					<th>Cluster</th>
+					<th>Replica</th>
+					<th>Elected Time</th>
+					<th>Time Until Update</th>
+					<th>Time Until Failover</th>
+				</tr>
+			</thead>
+			<tbody>
+				{{ range .Elected }}
+				<tr>
+					<td>{{ .UserID }}</td>
+					<td>{{ .Cluster }}</td>
+					<td>{{ .Replica }}</td>
+					<td>{{ .ElectedAt }}</td>
+					<td>{{ .UpdateTime }}</td>
+					<td>{{ .FailoverTime }}</td>
+				</tr>
+				{{ end }}
+			</tbody>
+		</table>
+	</body>
+</html>`
+
+var trackerTmpl *template.Template
+
+func init() {
+	trackerTmpl = template.Must(template.New("ha-tracker").Parse(trackerTpl))
+}
+
+func (h *haTracker) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	h.electedLock.RLock()
+	defer h.electedLock.RUnlock()
```

Comment:
Can we unlock right after we use it? I see we hold this lock until the network transfer is over and I see it as a potential DOS target, with someone deliberately holding the connection, as holding the lock will mean requests fail.